src/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs

```diff
@@ -389,9 +395,11 @@ private void BoundsCheck(uint sliceStartIndex, object? sliceStartObject, uint sl
                 // Multi-Segment Sequence
                 // This optimization works because we know sliceStartIndex, sliceEndIndex, startIndex, and endIndex are all >= 0
                 Debug.Assert(sliceStartIndex >= 0 && startIndex >= 0 && endIndex >= 0);
+                long startRunningIndex = ((ReadOnlySequenceSegment<T>?)sliceStartObject)?.RunningIndex ?? 0;
```

Comment:
Let's follow the similar pattern as the above BoundsCheck method and do an explicit check. It produces smaller assembly (we save ~8 instructions).

[sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5ilwDcNGsQDMTUgwDCDAN40GipuOAQIAGwYBZcgAoIwAFYwwGBrnUBLMDADKGbFAwB5IyYxoGr46fNWbAKIAdgAm3u6eHJZBZhbWdg5OAJKhMAiR0bH+MMEhKSFpAJRyCkpl6hBBAObmiRgAShxBQdFV+WkMALwMurr1MNhhQeoAnrYwAI4cMEE241X4MxgAPMAjGDAAfIVxc3XhpoUA/CyNza3tCAxHRwx0QtRl5ZU1MyFnLdWXXT19A0OjcZTGZzGALJardZbHbZXIHDDHU5NT5tVJXG53B5PJSlbEcCrVPzxeyOBrYaowH66fEvQq6XB1D4XNEMADURL2pMuhSx2IYNMJuxyoXq5KqlO61IJVTpbyZXxZ7KFuW5vOxuKexAA7ByEqTRRSGJtdbkDeK1QwAL4ajViBiqDTaUj6Ny+IUkpzwzzwk2hL38zK6j0YS4ZGK+vJo4ryR586W1fXI5kFK7de4a56EuVJhUpn7p2PqwtPSxcHru/auswAQm6QXx6kKGaUMb5TwZifOuY6kr+g2cwzGk2mszsYMWMUhG22FdJ8MKSK7qJTFrK1uLZWbilL5dhfqrDFrDHr6kbW5KG752aX317/X7g6BI9B4Mna2nMPicKrC/ly7Sq5KOubbngKNSzk4ZoSj0YF0h2kE5v+VxKtkwaqqB8bKiKYrQVKtK6NeKLfChX6hOhl6KOe2pBoyOFGhGUEWuulpAA==)

Something like:
```C#
        long startRunningIndex = 0;
        long endRunningIndex = 0;
        
        if (sliceStartObject != null)
        {
            startRunningIndex = ((ReadOnlySequenceSegment<byte>)sliceStartObject).RunningIndex;
        }
        
        if (sliceEndObject != null)
        {
            endRunningIndex = ((ReadOnlySequenceSegment<byte>)sliceEndObject).RunningIndex;
        }
```

We could even try removing the `startRunningIndex`/`endRunningIndex` locals all together (and maybe save ~6 more instructions), if they are not needed elsewhere and don't mess up the additions below (with overflow in mind):
[2nd sharlab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5ilwDcNGsQDMTUgwDCDAN40GipuOAQIAGwYBZcgAoIwAFYwwGBrnUBLMDADKGbFAwB5IyYxoGr46fNWbAKIAdgAm3u6eHJZBZhbWdg5OAJKhMAiR0bH+MMEhKSFpAJRyCkpl6hBBAObmiRgAShxBQdFV+WkMALwMurr1MNhhQeoAnrYwAI4cMEE241X4MxgAPMAjGDAAfIVxc3XhpoUA/CyNza3tCAxHRwx0QtRl5ZU1MyFnLdWXXT19A0OjcZTGZzGALJardZbHbZXIHDDHU5NT5tVJXG53B5PJSlbEcCrVPzxeyOBrYaowH66fEvQq6XB1D4XNEMADURL2pMuhSx2IYNMJuxyoXq5KqlO61IJVTpbyZXxZ7KFuW5vOxuKexAA7ByEqTRRSGJtdbkDeK1QwAL4ajViBiqDTaUj6Ny+IUkpzwzzwk2hL38zK6j0YS4ZGK+vJo4ryR586W1fXI5kFK7de4a56EuVJhUpn7p2PqwtPSxcHru/auswAQm6QXx6kKGaUMb5TwZifOuY6kr+g2cwzGk2mszsYMWMUhG22FdJ8MKSK7qJTFrK1uLZWbilL5dhfqrDFrDHr6kbW5KG752aX317/X7g6BI9B4Mna2nMPicKrC/ly7Sq5KOubbngKNSzk4ZoSj0YF0h2kE5v+VxKtkwaqqB8bKiKYrQVKtK6NeKLfChX6hOhl6KOe2pBoyOFGhGUEWsBOLFnaDqaFooguj4WTEpWPHegeWFhFWYa8ZyyRomJEbchebZgTR+p0d0EEhmigGKApwlQT8wmXBpDDnju9Kofx7iHnWDZNhRcltkoqk6aykqwb094AkOwKjvME4rO+0KqfOi5Eep57MUWbbGcJPpHieZ42a2dmKNpdFOTB0p0n27lPiCY6vr5UIznuIk8b+iH6aF55UTqDl0cayUUkxNCWkAA==)